proposals/p5337.md

```diff
@@ -789,16 +790,57 @@ final match_first {
 fn G(U:! Y) {
   // Can assume `U impls J` and `U.(J.A)` is `bool`.
 }
+
+fn H(V:! W) {
+  // Don't know if `V impls Y`, so we don't know
+  // which `impl` is selected. We use a symbolic
+  // witness for `V` and `V.(J.A)` is unknown
+  // (despite that it would be `bool` in either
+  // case).
+}
 ```
 
-> FIXME _**QUESTION**_: We could say that in this situation the value of `A.J`
-> is known to be `bool` if the query matches `[V:! W] V as J`, even though that
-> `impl` may not be used. The logic is that if that impl is not used, the
-> `impl forall [U:! Y] U as J` will be selected instead and it sets `A.J` to the
-> same thing. This is a bit like what we did in
-> [proposal #2868](https://github.com/carbon-language/carbon-lang/pull/2868)
-> which allowed an `impl` to overlap a `final impl` as long as agreed on the
-> overlap.
+The compiler will only use a known matching final `impl`'s associated constants
+if it is the first that _could_ match, as determined by type structure. For
```

Comment:
```suggestion
if it is the first that _could_ match, as determined by matching the query type structure against each `impl` in the `match_first`. For
```

I don't know if we have a designated verb for what we do with the type structure - I have been calling it matching, so suggested that here.

The function in the code is `IsCompatibleWith()` so maybe we don't have a verb and we use a relational adjective instead.

```suggestion
if it is the first that _could_ match, as determined by the query type structure being found to be compatible or not with each `impl` in the `match_first`. For
```

Reply:
> I don't think your text makes sense, since only `impl`s have type structure, not queries. I'll try a rephrase.

I guess, in the design. In the implementation, queries are type structures built from the self type and each interface in the facet type query.

> Note though I suspect the function IsCompatibleWith does something different, since we want to consider a `[V:! W] V as L(bool)` query as "could match" `impl forall [U:! Y] C(U) as L(bool)` even though just looking at the type structure it won't consider it matching.

These match via IsCompatiableWIth. The symbolic `V` consumes the `C(` as well as everything to to the matching `)` so they are compatible.